tools/pythonpkg/src/pandas/analyzer.cpp

```diff
@@ -111,38 +117,135 @@ static LogicalType ConvertStructToMap(LogicalType &map_value_type) {
 	return LogicalType::MAP(LogicalType::VARCHAR, map_value_type);
 }
 
+// This is similar to ForceMaxLogicalType but we have custom rules around combining STRUCT types
+// And because of that we have to avoid ForceMaxLogicalType for every nested type
 static bool UpgradeType(LogicalType &left, const LogicalType &right) {
-	bool compatible = CheckTypeCompatibility(left, right);
-	if (!compatible) {
-		return false;
+	if (left.id() == LogicalTypeId::SQLNULL) {
+		// Early out for upgrading null
+		left = right;
+		return true;
 	}
-	// If struct constraints are not respected, left will be set to MAP
-	if (left.id() == LogicalTypeId::STRUCT && right.id() == left.id()) {
-		bool valid_struct = IsStructColumnValid(left, right);
-		if (valid_struct) {
-			child_list_t<LogicalType> children;
-			for (idx_t i = 0; i < StructType::GetChildCount(right); i++) {
-				auto &right_child = StructType::GetChildType(right, i);
-				auto new_child = StructType::GetChildType(left, i);
-				auto child_name = StructType::GetChildName(left, i);
-				if (!UpgradeType(new_child, right_child)) {
+
+	if (left.IsNested() && right.id() == LogicalTypeId::SQLNULL) {
+		return true;
+	}
+
+	switch (left.id()) {
+	case LogicalTypeId::LIST: {
+		if (right.id() != left.id()) {
+			// Not both sides are LIST, not compatible
+			// FIXME: maybe compatible with ARRAY type??
+			return false;
```

Comment:
Can we add tests that trigger some of these different (new) behaviors?

Reply:
I'm not really sure what to add tests for, I believe there is no new behavior here, just a bug in the handling of structs in nested types.

Previously everything was handled with `left = LogicalType::ForceMaxLogicalType`, but that has different rules concerning structs.

```
[
    {'key1': 42}
],
[
    {'key1': 42, 'key2': 42}
]
```

individually these become `STRUCT(key1 integer)[]` and `STRUCT(key1 integer, key2 integer)[]`
When we use ForceMaxLogicalType this just returns whatever is `left` because the types are not considered compatible.

In UpgradeType we turn cases like this into a `MAP(VARCHAR, INTEGER)`

```
[
    {'key1': 42},
    {'key1': 42, 'key2': 42}
]
```

The change that this PR makes is that when combining the previous example, this same behavior now applies

Reply:
This PR is only as big as it is because all the other nested types can no longer be handled by `ForceMaxLogicalType` because they *could* contain a STRUCT

Reply:
More tests with nested types with structs in it please